google/cloud/internal/timer_queue.cc

```diff
@@ -46,80 +50,80 @@ void TimerQueue::Shutdown() {
-             HasExpiredTimer(lk, std::chrono::system_clock::now());
+  // The is_leader flag allows us to restart this loop without worrying about
+  // electing new leaders.
+  auto is_leader = false;
+  std::unique_lock<std::mutex> lk(mu_);
+  while (!shutdown_) {
+    if (!is_leader && has_leader_) {
+      // The current thread becomes a follower while there is a leader.
+      cv_follower_.wait(lk, [this] { return shutdown_ || !has_leader_; });
+      continue;
+    }
+    is_leader = true;
+    has_leader_ = true;
+    if (timers_.empty()) {
+      cv_.wait(lk, [this] { return shutdown_ || !timers_.empty(); });
+      continue;
+    }
+    // Should a new timer appear that changes the "first" timer, we need to
+    // wake up and recompute the sleep time. But note that the leader thread
+    // does not need to relinquish its role to do so.
+    auto until = timers_.begin()->first;
+    auto predicate = [this, until] {
+      return shutdown_ || timers_.empty() || timers_.begin()->first != until;
     };
-    // TODO(#10512): Only wake one thread to service the next timer instead
-    // of all of them.
-    cv_.wait_until(lk, until, predicate);
-    if (shutdown_) return NotifyShutdown(std::move(lk));
-    auto now = std::chrono::system_clock::now();
-    if (HasExpiredTimer(lk, now)) {
-      ExpireTimer(std::move(lk), now);
+    if (cv_.wait_until(lk, until, std::move(predicate))) {
+      // Timers can be expired only if wait_util() exits due to a timeout.
+      continue;
     }
+    // if we get here we know that predicate() is false, which implies that
+    // `timers_` is not empty and the first timer's key is `until`.
+    auto p = std::move(timers_.begin()->second);
+    timers_.erase(timers_.begin());
+    // Relinquish the leader role, release the mutex and then signal a follower.
+    has_leader_ = false;
+    lk.unlock();
+    is_leader = false;
+    // Elect a new leader (if available) to continue expiring timers.
+    cv_follower_.notify_one();
+    // This may run user code (in the continuations for the future). That may
+    // do all kinds of things, including calling back into this class to create
+    // new timers. We cannot hold the mutex while it is running.
+    p.set_value(until);
+    lk.lock();
   }
+  CancelAll(std::move(lk), "TimerQueue shutdown");
 }
 
 void TimerQueue::CancelAll() {
-  std::multimap<std::chrono::system_clock::time_point, PromiseType> timers;
-  {
-    std::lock_guard<std::mutex> lk(mu_);
-    std::swap(timers, timers_);
-  }
-  for (auto& timer : timers) {
-    timer.second.set_value(Status(StatusCode::kCancelled, "Timer cancelled"));
-  }
-}
-
-// Requires a lock to be held before calling.
-bool TimerQueue::HasExpiredTimer(std::unique_lock<std::mutex> const&,
-                                 std::chrono::system_clock::time_point tp) {
-  return !timers_.empty() && timers_.begin()->first < tp;
+  CancelAll(std::unique_lock<std::mutex>(mu_), "TimerQueue cancel all");
```

Comment:
It looks like this guy should then also `cv_.notify_all()` now that `timers_` has changed.

Reply:
I think there is at most one thread blocked on `cv_`.  `cv_.notify_one()` seems like enough, but probably won't make a difference.

In any case, I am planning to remove `CancelAll()` in favor of per-timer cancels.

Reply:
> I think there is at most one thread blocked on `cv_`. `cv_.notify_one()` seems like enough

I agree ... just the leader.  And therefore, that `Shutdown()` now need only `cv_.notify_one()` too.

> In any case, I am planning to remove `CancelAll()` in favor of per-timer cancels.

+1 ... It seemed difficult to use at best.